man/dminfo.1
.TH "DMINFO" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dminfo \(en outputs system and database information
.SH SYNOPSIS
\f[B]dminfo\f[R] \-\-help
.PP
\f[B]dminfo\f[R] \-\-version
.PP
\f[B]dminfo\f[R] [\-\-\f[B]database\f[R] \f[I]file\f[R]]
.SH DESCRIPTION
The \f[B]dminfo\f[R] utility program prints build, database, and system
information to standard output.
The path to the beat, log, or observation database is passed through
command\-line argument \f[CR]\-\-database\f[R].
.PP
The output contains compiler version and options; database PRAGMAs,
tables, and number of rows; as well as system name, version, and host
name.
.SH OPTIONS
.TP
\f[B]\-\-database\f[R], \f[B]\-d\f[R] \f[I]file\f[R]
File path to SQLite observation database.
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Information has been printed.
.TP
\f[B]1\f[R]
Failure.
Program execution failed.
.SH EXAMPLE
Print build, database, and system information:
.IP
.EX
$ dminfo \-d /var/dmpack/observ.db
.EE


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema
* `Changed` refactored command-line argument parsing
* `Changed` refactored logical function names
* `Fixed` wrong baud rate in GeoCOM class

### Programs

* `Added` program `dmbot` for remote control of clients through XMPP
* `Added` program `dmdwd` for Deutscher Wetterdienst (DWD) API monitoring
* `Added` program `dmmb` for sensor monitoring through Modbus RTU/TCP
* `Added` program `dmmbctl` for Modbus RTU/TCP command-line control
* `Added` program `dmsystem` for system monitoring
* `Added` program `dmved` for MPPT and battery monitoring (VE.Direct)
* `Added` interactive map view to `dmweb` (Leaflet)

### Documentation


app/dminfo.f90
! dminfo.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dminfo
    !! Utility program that displays system and database information.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dminfo'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    type :: app_type
        !! Command-line arguments.
        character(FILE_PATH_LEN) :: database = ' ' !! Path to database (optional).
    end type app_type

    integer        :: rc  ! Return code.
    type(app_type) :: app ! App settings.

    call dm_init()

    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    call output_info(app)
contains
    subroutine output_info(app)
        !! Reads system and database information and prints it as key-value
        !! pairs to standard output.
        use, intrinsic :: iso_fortran_env, only: compiler_options, compiler_version

        character(*), parameter :: FALSE = 'false'
        character(*), parameter :: TRUE  = 'true'

        type(app_type), intent(inout) :: app

        integer     :: app_id, capacity, mode, ncore, rc, schema_version
        integer(i8) :: available, n, nbyte
        logical     :: foreign_keys, has

        character(64)             :: file_system, model, mounted_on
        character(:), allocatable :: mode_name
        type(db_type)             :: db
        type(posix_uname_type)    :: uname

        ! Database information.
        if (dm_string_has(app%database)) then
            rc = dm_db_open(db, app%database, read_only=.true.)
            if (dm_is_error(rc)) call dm_error_out(rc, 'failed to open database ' // app%database, fatal=.true.)

            rc = dm_db_get_application_id(db, app_id)
            rc = dm_db_get_foreign_keys(db, foreign_keys)
            rc = dm_db_get_journal_mode(db, mode, mode_name)
            rc = dm_db_get_schema_version(db, schema_version)
            rc = dm_db_size(db, nbyte)

            ! Available disk space and disk capacity.
            rc = dm_posix_disk_free(path        = app%database, &
                                    file_system = file_system,  &
                                    available   = available,    &
                                    capacity    = capacity,     &
                                    mounted_on  = mounted_on)

            print '("db.application_id: ", z0)', app_id
            print '("db.foreign_keys: ", a)',    dm_btoa(foreign_keys, TRUE, FALSE)
            print '("db.fs.available: ", i0)',   available
            print '("db.fs.capacity: ", i0)',    capacity
            print '("db.fs.mount_point: ", a)',  trim(mounted_on)
            print '("db.fs.path: ", a)',         trim(file_system)
            print '("db.journal_mode: ", a)',    mode_name
            print '("db.library: ", a)',         dm_db_version(.true.)
            print '("db.path: ", a)',            trim(app%database)
            print '("db.schema_version: ", i0)', schema_version
            print '("db.size: ", i0)',           nbyte

            has = dm_db_table_has(db, SQL_TABLE_BEATS)
            rc  = dm_db_count_beats(db, n)

            print '("db.table.beats: ", a)', dm_btoa(has, TRUE, FALSE)
            if (has) print '("db.table.beats.rows: ", i0)', n

            has = dm_db_table_has(db, SQL_TABLE_IMAGES)
            rc  = dm_db_count_images(db, n)

            print '("db.table.images: ", a)', dm_btoa(has, TRUE, FALSE)
            if (has) print '("db.table.images.rows: ", i0)', n

            has = dm_db_table_has(db, SQL_TABLE_LOGS)
            rc  = dm_db_count_logs(db, n)

            print '("db.table.logs: ", a)', dm_btoa(has, TRUE, FALSE)
            if (has) print '("db.table.logs.rows: ", i0)', n

            has = dm_db_table_has(db, SQL_TABLE_NODES)
            rc  = dm_db_count_nodes(db, n)

            print '("db.table.nodes: ", a)', dm_btoa(has, TRUE, FALSE)
            if (has) print '("db.table.nodes.rows: ", i0)', n

            has = dm_db_table_has(db, SQL_TABLE_OBSERVS)
            rc  = dm_db_count_observs(db, n)

            print '("db.table.observs: ", a)', dm_btoa(has, TRUE, FALSE)
            if (has) print '("db.table.observs.rows: ", i0)', n

            has = dm_db_table_has(db, SQL_TABLE_RECEIVERS)
            rc  = dm_db_count_receivers(db, n)

            print '("db.table.receivers: ", a)', dm_btoa(has, TRUE, FALSE)
            if (has) print '("db.table.receivers.rows: ", i0)', n

            has = dm_db_table_has(db, SQL_TABLE_RESPONSES)
            rc  = dm_db_count_responses(db, n)

            print '("db.table.responses: ", a)', dm_btoa(has, TRUE, FALSE)
            if (has) print '("db.table.responses.rows: ", i0)', n

            has = dm_db_table_has(db, SQL_TABLE_SENSORS)
            rc  = dm_db_count_sensors(db, n)

            print '("db.table.sensors: ", a)', dm_btoa(has, TRUE, FALSE)
            if (has) print '("db.table.sensors.rows: ", i0)', n

            has = dm_db_table_has(db, SQL_TABLE_TARGETS)
            rc  = dm_db_count_targets(db, n)

            print '("db.table.targets: ", a)', dm_btoa(has, TRUE, FALSE)
            if (has) print '("db.table.targets.rows: ", i0)', n

            has = dm_db_table_has(db, SQL_TABLE_TRANSFERS)
            rc  = dm_db_count_transfers(db, n)

            print '("db.table.transfers: ", a)', dm_btoa(has, TRUE, FALSE)
            if (has) print '("db.table.transfers.rows: ", i0)', n

            call dm_db_close(db)
        end if

        call dm_posix_uname(uname)
        rc = dm_posix_cpu_cores(ncore)
        rc = dm_posix_cpu_model(model)

        print '("dmpack.compiler: ", a)',   compiler_version()
        print '("dmpack.date: ", a)',       DM_BUILD_DATE
        print '("dmpack.options: ", a)',    compiler_options()
        print '("dmpack.version: ", a)',    DM_VERSION_STRING
        print '("system.byte_order: ", a)', dm_btoa(LITTLE_ENDIAN, 'little-endian', 'big-endian')
        print '("system.cpu.cores: ", i0)', ncore
        print '("system.cpu.model: ", a)',  trim(model)
        print '("system.hostname: ", a)',   trim(uname%node_name)
        print '("system.name: ", a)',       trim(uname%system_name)
        print '("system.platform: ", a)',   trim(uname%machine)
        print '("system.release: ", a)',    trim(uname%release)
        print '("system.time.now: ", a)',   dm_time_now()
        print '("system.time.zone: ", a)',  dm_time_zone()
        print '("system.version: ", a)',    trim(uname%version)
    end subroutine output_info

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments.
        type(app_type), intent(out) :: app

        type(arg_parser_class) :: parser

        call parser%add(name='database', short='d', type=ARG_TYPE_DATABASE, exist=.true.) ! -d, --database <path>
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return
        call parser%get('database', app%database)
    end function read_args

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a)', dm_db_version(.true.)
    end subroutine version_callback
end program dminfo
